### ISSUE TITLE
5.8.0-RC2 segfault (combination of DESTROY, goto, delete)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5679]
Migrated from [rt.perl.org#9990](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=9990) (status was 'resolved')

Searchable as RT9990$

**Comment by p5pRT** [comment]
### From pixel@mandrakesoft.com



The following code segfaults&#8203;:

sub A&#8203;::DESTROY \{
&nbsp;    my $f = delete $A&#8203;::list\{$\_\[0\]\};
&nbsp;    goto $f;
\}

\{
&nbsp;    my $w;
&nbsp;    my $b = bless \{\}\, 'A';
&nbsp;    $A&#8203;::list\{$b\} = sub \{ $w \};
&nbsp;    0;
\}

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by pixel at Tue Jun 25 17:11:59 CEST 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.18-10mdksmp, archname=i386-linux
    uname='linux no.mandrakesoft.com 2.4.18-10mdksmp #1 smp wed apr 10 13:53:12 cest 2002 i686 unknown '
    config_args='-des -Darchname=i386-linux -Dcc=gcc -Doptimize=-O3 -fomit-frame-pointer -pipe -mcpu=pentiumpro -march=i586 -ffast-math -fno-strength-reduce -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr -Dman3ext=3pm -Dcf_by=MandrakeSoft -Dmyhostname=localhost -Dperladmin=root@localhost -Dd_dosuid -Ud_csh -Duseshrplib'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O3 -fomit-frame-pointer -pipe -mcpu=pentiumpro -march=i586 -ffast-math -fno-strength-reduce',
    cppflags='-fno-strict-aliasing -I/usr/include/gdbm'
    ccversion='', gccversion='3.1.1 (Mandrake Linux 8.3 3.1.1-0.6mdk)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version='2.2.5'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic -Wl,-rpath,/usr/lib/perl5/5.8.0/i386-linux/CORE'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /home/pixel/etc/perl/i386-linux
    /home/pixel/etc/perl
    /home/pixel/etc/perl/i386-linux
    /home/pixel/etc/perl/site_perl/i386-linux
    /home/pixel/etc/perl/site_perl
    /usr/lib/perl5/5.8.0/i386-linux
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl/5.6.1
    /usr/lib/perl5/site_perl/5.6.0
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i386-linux
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    .


Environment for perl v5.8.0:
    HOME=/home/pixel
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/home/pixel/lib/i386
    LOGDIR (unset)
    PATH=/home/pixel/bin:/home/pixel/bin/oneliner:/home/pixel/bin/perl:/home/pixel/bin/ruby:/home/pixel/bin/python:/home/pixel/bin/tcl:/home/pixel/bin/i386:/usr/freeware/bin:/usr/etc:/usr/local/bin:/usr/X11R6/bin:/sbin:/usr/sbin:/usr/X11R6/bin:/bin:/usr/bin:/usr/local/bin:/root/bin:/root/bin:/usr/X11R6/bin
    PERL5LIB=/home/pixel/etc/perl:/home/pixel/etc/perl/i386-linux:/home/pixel/etc/perl/site_perl
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @lizmat

At 01&#8203;:14 AM 6/26/02 \+0200\, Pixel wrote&#8203;:
>The following code segfaults&#8203;:
>
>sub A&#8203;::DESTROY \{
>     my $f = delete $A&#8203;::list\{$\_\[0\]\};
>     goto $f;
>\}
>
>\{
>     my $w;
>     my $b = bless \{\}\, 'A';
>     $A&#8203;::list\{$b\} = sub \{ $w \};
>     0;
>\}

Doesn't happen on 5\.6\.0\.

Note that removing the $w from the sub does not segfault\.  So it has 
something to do with the closure that was created\.

Also note that combining the two statements of DESTROY into&#8203;:

&nbsp;   goto delete $A&#8203;::list\{$\_\[0\]\};

also does not segfault\.  So it would seem that the creation of the lexicals 
in the DESTROY is interfering with the closure\.


Just my 2c worth\.\.\.


Liz

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]


On Wednesday\, June 26\, 2002\, at 03&#8203;:08  PM\, Elizabeth Mattijsen wrote&#8203;:

> At 01&#8203;:14 AM 6/26/02 \+0200\, Pixel wrote&#8203;:
>> The following code segfaults&#8203;:
>>
>> sub A&#8203;::DESTROY \{
>>     my $f = delete $A&#8203;::list\{$\_\[0\]\};
>>     goto $f;
>> \}
>>
>> \{
>>     my $w;
>>     my $b = bless \{\}\, 'A';
>>     $A&#8203;::list\{$b\} = sub \{ $w \};
>>     0;
>> \}
>
>
>
> Just my 2c worth\.\.\.

Then we've got 4c by now\.\.\.;\-\)


Here's a bt&#8203;:

Thread 0 Crashed&#8203;:
&nbsp;  \#0   0x000000fc in 0xfc
&nbsp;  \#1   0x00430844 in Perl\_runops\_debug
&nbsp;  \#2   0x003c0a80 in S\_call\_body
&nbsp;  \#3   0x003c05b8 in Perl\_call\_sv
&nbsp;  \#4   0x0046c45c in Perl\_sv\_clear
&nbsp;  \#5   0x0046cff4 in Perl\_sv\_free
&nbsp;  \#6   0x004734fc in Perl\_sv\_unref\_flags
&nbsp;  \#7   0x00469e20 in Perl\_sv\_force\_normal\_flags
&nbsp;  \#8   0x0049f4b0 in Perl\_leave\_scope
&nbsp;  \#9   0x0049b674 in Perl\_pop\_scope
&nbsp;  \#10  0x004aa86c in Perl\_pp\_leaveloop
&nbsp;  \#11  0x00430844 in Perl\_runops\_debug
&nbsp;  \#12  0x003bf7d8 in S\_run\_body
&nbsp;  \#13  0x003bf19c in perl\_run
&nbsp;  \#14  0x00002678 in main
&nbsp;  \#15  0x00002298 in \_start
&nbsp;  \#16  0x000020c8 in start


It appears that pp\_ctl\.c&#8203;:2162 has&#8203;:

\(void\)\(\*CvXSUB\(cv\)\)\(aTHX\_ cv\);

which comes out to a call&#8203;:
\(\(XPVCV\*\)SvANY\(sv\)\)\->xcv\_xsub

xcv\_xsub seems to be 0xFF for some reason\. \(at least on two subsequent 
tries\.\.\.\)


If anyone has more tuits\.\.\.\.good luck\.


regards\,

kay

**Comment by p5pRT** [comment]
### From @iabyn

Bug \#9990 demonstrates a coredump\, which can be reduced to the following&#8203;:

&nbsp;    sub f \{ my $x; my $a = sub \{ $x \}; goto $a \} f\(\);

Basically the goto first unwinds f's scope\, which causes $a to be freed;
then it tries to call &$a\. Bang\.\.\.\.\.

This patch makes pp\_goto add an extra ref to the CV for the duration\.

\-\- 
You live and learn \(although usually you just live\)\.


\# This is a patch for 19021\.ORIG to update it to 19021
\# 
\# To apply this patch&#8203;:
\# STEP 1&#8203;: Chdir to the source directory\.
\# STEP 2&#8203;: Run the 'applypatch' program with this patch file as input\.
\#
\# If you do not have 'applypatch'\, it is part of the 'makepatch' package
\# that you can fetch from the Comprehensive Perl Archive Network&#8203;:
\# http&#8203;://www\.perl\.com/CPAN/authors/Johan\_Vromans/makepatch\-x\.y\.tar\.gz
\# In the above URL\, 'x' should be 2 or higher\.
\#
\# To apply this patch without the use of 'applypatch'&#8203;:
\# STEP 1&#8203;: Chdir to the source directory\.
\# STEP 2&#8203;: Run the 'patch' program with this file as input\.
\#
\#\#\#\# End of Preamble \#\#\#\#

\#\#\#\# Patch data follows \#\#\#\#
diff \-up '19021\.ORIG/pp\_ctl\.c' '19021/pp\_ctl\.c'
Index&#8203;: \./pp\_ctl\.c
<details><summary>Inline Patch</summary>

```diff
--- ./pp_ctl.c	Tue Mar 18 21:21:58 2003
+++ ./pp_ctl.c	Tue Mar 18 21:38:14 2003
@@ -2203,6 +2203,7 @@ PP(pp_goto)
 	    }
 
 	    /* First do some returnish stuff. */
+	    SvREFCNT_inc(cv); /* avoid premature free during unwind */
 	    FREETMPS;
 	    cxix = dopoptosub(cxstack_ix);
 	    if (cxix < 0)
@@ -2250,6 +2251,7 @@ PP(pp_goto)
 
 	    /* Now do some callish stuff. */
 	    SAVETMPS;
+	    SAVEFREESV(cv); /* later, undo the 'avoid premature free' hack */
 	    if (CvXSUB(cv)) {
 #ifdef PERL_XSUB_OLDSTYLE
 		if (CvOLDSTYLE(cv)) {
diff -up '19021.ORIG/t/op/goto.t' '19021/t/op/goto.t'

```
</details>
Index: ./t/op/goto.t
<details><summary>Inline Patch</summary>

```diff
--- ./t/op/goto.t	Tue Mar 18 21:22:18 2003
+++ ./t/op/goto.t	Tue Mar 18 21:28:24 2003
@@ -2,7 +2,7 @@
 
 # "This IS structured code.  It's just randomly structured."
 
-print "1..27\n";
+print "1..28\n";
 
 while ($?) {
     $foo = 1;
@@ -177,6 +177,14 @@ print ($ok ? "ok 22\n" : "not ok 22\n");
     print "ok 27 - weird case of goto and for(;;) loop\n";
 }
 
+# bug #9990 - don't prematurely free the CV we're &going to.
+
+sub f1 {
+    my $x;
+    goto sub { $x; print "ok 28 - don't prematurely free CV\n" }
+}
+f1();
+
 exit;
 
 bypass:

```
</details>
#### End of Patch data ####

\#\#\#\# ApplyPatch data follows \#\#\#\#
\# Data version        : 1\.0
\# Date generated      : Tue Mar 18 21&#8203;:40&#8203;:00 2003
\# Generated by        : makepatch 2\.00\_05
\# Recurse directories : Yes
\# Excluded files      : keywords\\\.h|warnings\\\.h|regnodes\\\.h|perlapi\\\.c|perlapi\\\.h|global\\\.sym|embedvar\\\.h|embed\\\.h|pod\\/perlapi\\\.pod|pod\\/perlintern\\\.pod|proto\\\.h
\# v 'patchlevel\.h' 4571 1048022560 33188
\# p 'pp\_ctl\.c' 87366 1048023494 0100644
\# p 't/op/goto\.t' 4052 1048022904 0100755
\#\#\#\# End of ApplyPatch data \#\#\#\#

\#\#\#\# End of Patch kit \[created&#8203;: Tue Mar 18 21&#8203;:40&#8203;:00 2003\] \#\#\#\#
\#\#\#\# Patch checksum&#8203;: 63 1851 1920 \#\#\#\#
\#\#\#\# Checksum&#8203;: 81 2526 57466 \#\#\#\#

**Comment by p5pRT** [comment]
### From @rgs

Dave Mitchell wrote&#8203;:
> Bug \#9990 demonstrates a coredump\, which can be reduced to the following&#8203;:
> 
>     sub f \{ my $x; my $a = sub \{ $x \}; goto $a \} f\(\);
> 
> Basically the goto first unwinds f's scope\, which causes $a to be freed;
> then it tries to call &$a\. Bang\.\.\.\.\.
> 
> This patch makes pp\_goto add an extra ref to the CV for the duration\.

Thanks\, applied as \#19025\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @nwc10

On Tue Jun 25 23&#8203;:14&#8203;:28 2002\, RT\_System wrote&#8203;:
> At 01&#8203;:14 AM 6/26/02 \+0200\, Pixel wrote&#8203;:
> >The following code segfaults&#8203;:
> >
> >sub A&#8203;::DESTROY \{
> >     my $f = delete $A&#8203;::list\{$\_\[0\]\};
> >     goto $f;
> >\}
> >
> >\{
> >     my $w;
> >     my $b = bless \{\}\, 'A';
> >     $A&#8203;::list\{$b\} = sub \{ $w \};
> >     0;
> >\}
> 
> Doesn't happen on 5\.6\.0\.

SEGV was caused as a side effect of


commit 840a7b70755d06740715e982aa756f9d77203c4e
Author&#8203;: Ilya Zakharevich \<ilya@&#8203;math\.berkeley\.edu>
Date&#8203;:   Mon Dec 4 19&#8203;:40&#8203;:25 2000 \-0500

&nbsp;    Re&#8203;: \[PATCH\] The largest hoax of all times?
&nbsp;    Date&#8203;: Tue\, 5 Dec 2000 00&#8203;:40&#8203;:25 \-0500
&nbsp;    Message\-ID&#8203;: \<20001205004025\.A4050@&#8203;monk\.mps\.ohio\-state\.edu>
&nbsp;    
&nbsp;    Subject&#8203;: Re&#8203;: \[PATCH\] The largest hoax of all times?
&nbsp;    From&#8203;: Ilya Zakharevich \<ilya@&#8203;math\.ohio\-state\.edu>
&nbsp;    Date&#8203;: Mon\, 4 Dec 2000 23&#8203;:55&#8203;:53 \-0500
&nbsp;    Message\-ID&#8203;: \<20001204235553\.A1140@&#8203;monk\.mps\.ohio\-state\.edu>
&nbsp;    
&nbsp;    Subject&#8203;: Re&#8203;: \[PATCH\] The largest hoax of all times?
&nbsp;    From&#8203;: Ilya Zakharevich \<ilya@&#8203;math\.ohio\-state\.edu>
&nbsp;    Date&#8203;: Tue\, 5 Dec 2000 01&#8203;:28&#8203;:45 \-0500
&nbsp;    Message\-ID&#8203;: \<20001205012844\.A4227@&#8203;monk\.mps\.ohio\-state\.edu>
&nbsp;    
&nbsp;    Fix the unpredictable order of DESTROYs\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;7991

\[this bug was fixed in 2003\, but it's related to \#99850\]

Nicholas Clark

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Tue Jun 25 23&#8203;:14&#8203;:28 2002\, RT\_System wrote&#8203;:
> At 01&#8203;:14 AM 6/26/02 \+0200\, Pixel wrote&#8203;:
> >The following code segfaults&#8203;:
> >
> >sub A&#8203;::DESTROY \{
> >     my $f = delete $A&#8203;::list\{$\_\[0\]\};
> >     goto $f;
> >\}
> >
> >\{
> >     my $w;
> >     my $b = bless \{\}\, 'A';
> >     $A&#8203;::list\{$b\} = sub \{ $w \};
> >     0;
> >\}
> 
> Doesn't happen on 5\.6\.0\.

SEGV was caused as a side effect of


commit 840a7b70755d06740715e982aa756f9d77203c4e
Author&#8203;: Ilya Zakharevich \<ilya@&#8203;math\.berkeley\.edu>
Date&#8203;:   Mon Dec 4 19&#8203;:40&#8203;:25 2000 \-0500

&nbsp;    Re&#8203;: \[PATCH\] The largest hoax of all times?
&nbsp;    Date&#8203;: Tue\, 5 Dec 2000 00&#8203;:40&#8203;:25 \-0500
&nbsp;    Message\-ID&#8203;: \<20001205004025\.A4050@&#8203;monk\.mps\.ohio\-state\.edu>
&nbsp;    
&nbsp;    Subject&#8203;: Re&#8203;: \[PATCH\] The largest hoax of all times?
&nbsp;    From&#8203;: Ilya Zakharevich \<ilya@&#8203;math\.ohio\-state\.edu>
&nbsp;    Date&#8203;: Mon\, 4 Dec 2000 23&#8203;:55&#8203;:53 \-0500
&nbsp;    Message\-ID&#8203;: \<20001204235553\.A1140@&#8203;monk\.mps\.ohio\-state\.edu>
&nbsp;    
&nbsp;    Subject&#8203;: Re&#8203;: \[PATCH\] The largest hoax of all times?
&nbsp;    From&#8203;: Ilya Zakharevich \<ilya@&#8203;math\.ohio\-state\.edu>
&nbsp;    Date&#8203;: Tue\, 5 Dec 2000 01&#8203;:28&#8203;:45 \-0500
&nbsp;    Message\-ID&#8203;: \<20001205012844\.A4227@&#8203;monk\.mps\.ohio\-state\.edu>
&nbsp;    
&nbsp;    Fix the unpredictable order of DESTROYs\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;7991

\[this bug was fixed in 2003\, but it's related to \#99850\]

Nicholas Clark
